libnavigation-core/src/test/java/com/mapbox/navigation/core/internal/HistoryRecordingStateHandlerTest.kt

```diff
@@ -0,0 +1,753 @@
+            observer2.onShouldStopRecording(any<NavigationSessionState.ActiveGuidance>())
+            observer2.onShouldStartRecording(any<NavigationSessionState.FreeDrive>())
+            observer3.onShouldStopRecording(any<NavigationSessionState.ActiveGuidance>())
+            observer3.onShouldStartRecording(any<NavigationSessionState.FreeDrive>())
+        }
+
+        // 2 observers
+        clearMocks(observer1, observer2, observer3)
+
+        stateHandler.unregisterStateChangeObserver(observer2)
+        stateHandler.setRoutes(listOf(mockk()))
+
+        verify(exactly = 1) {
+            observer1.onShouldStopRecording(any<NavigationSessionState.FreeDrive>())
+            observer1.onShouldStartRecording(any<NavigationSessionState.ActiveGuidance>())
+            observer3.onShouldStopRecording(any<NavigationSessionState.FreeDrive>())
+            observer3.onShouldStartRecording(any<NavigationSessionState.ActiveGuidance>())
+        }
+        verifyNoInteractions(observer2)
+
+        // no observers
+        clearMocks(observer1, observer2, observer3)
+
+        stateHandler.unregisterAllStateChangeObservers()
+        stateHandler.setRoutes(emptyList())
+
+        verifyNoInteractions(observer1, observer2, observer3)
+    }
+
+    @Test
+    fun setEmptyRoutesOverEmptyRoutesNotDriving() {
+        stateHandler.setRoutes(emptyList())
+
+        verifyNoInteractions(observer)
+    }
+
+    @Test
+    fun setEmptyRoutesOverEmptyRoutesDriving() {
+        stateHandler.onSessionStateChanged(TripSessionState.STARTED)
+        clearMocks(observer)
+
+        stateHandler.setRoutes(emptyList())
+
+        verifyNoInteractions(observer)
+    }
+
+    @Test
+    fun setEmptyRoutesOverEmptyRoutesStoppedDriving() {
+        stateHandler.onSessionStateChanged(TripSessionState.STARTED)
+        stateHandler.onSessionStateChanged(TripSessionState.STOPPED)
+        clearMocks(observer)
+
+        stateHandler.setRoutes(emptyList())
+
+        verifyNoInteractions(observer)
+    }
+
+    @Test
+    fun setEmptyRoutesOverEmptyRoutesFailed() {
+        stateHandler.setRoutes(listOf(mockk()))
+        stateHandler.lastSetRoutesFailed()
+        clearMocks(observer)
+
+        stateHandler.setRoutes(emptyList())
+
+        verifyNoInteractions(observer)
+    }
+
+    @Test
+    fun setNonEmptyRoutesOverEmptyRoutesNotDriving() {
+        stateHandler.setRoutes(listOf(mockk()))
+
+        verifyNoInteractions(observer)
+    }
+
+    @Test
+    fun setNonEmptyRoutesOverEmptyRoutesDriving() {
+        stateHandler.onSessionStateChanged(TripSessionState.STARTED)
+
+        stateHandler.setRoutes(listOf(mockk()))
+
+        val freeDrives = mutableListOf<NavigationSessionState.FreeDrive>()
+        verify(exactly = 1) {
+            observer.onShouldStartRecording(capture(freeDrives))
+        }
+        verify(exactly = 1) {
+            observer.onShouldStopRecording(freeDrives[0])
+            observer.onShouldStartRecording(match(isActiveGuidance))
```

Comment:
```suggestion
            observer.onShouldStartRecording(ofType<NavigationSessionState.ActiveGuidance>())
```
The same below and for `isFreeDrive`.